lightning-persister/src/lib.rs

```diff
@@ -0,0 +1,395 @@
+				}
+				persisted_chan_data_1 = persister_1.load_channel_data::<EnforcingChannelKeys>().unwrap();
+				assert_eq!(persisted_chan_data_1.keys().len(), 1);
+				for mon in persisted_chan_data_1.values() {
+					assert_eq!(mon.get_latest_update_id(), $expected_update_id);
+				}
+			}
+		}
+
+		// Create some initial channel and check that a channel was persisted.
+		let _ = create_announced_chan_between_nodes(&nodes, 0, 1, InitFeatures::known(), InitFeatures::known());
+		check_persisted_data!(0);
+
+		// Send a few payments and make sure the monitors are updated to the latest.
+		send_payment(&nodes[0], &vec!(&nodes[1])[..], 8000000, 8_000_000);
+		check_persisted_data!(5);
+		send_payment(&nodes[1], &vec!(&nodes[0])[..], 4000000, 4_000_000);
+		check_persisted_data!(10);
+
+		// Force close because cooperative close doesn't result in any persisted
+		// updates.
+		nodes[0].node.force_close_channel(&nodes[0].node.list_channels()[0].channel_id);
+		check_closed_broadcast!(nodes[0], false);
+		check_added_monitors!(nodes[0], 1);
+
+		let node_txn = nodes[0].tx_broadcaster.txn_broadcasted.lock().unwrap();
+		assert_eq!(node_txn.len(), 1);
+
+		let header = BlockHeader { version: 0x20000000, prev_blockhash: Default::default(), merkle_root: Default::default(), time: 42, bits: 42, nonce: 42 };
+		connect_block(&nodes[1], &Block { header, txdata: vec![node_txn[0].clone(), node_txn[0].clone()]}, 1);
+		check_closed_broadcast!(nodes[1], false);
+		check_added_monitors!(nodes[1], 1);
+
+		// Make sure everything is persisted as expected after close.
+		check_persisted_data!(11);
+	}
+
+	// Windows ignores the read-only flag for folders.
+	#[cfg(not(target_os = "windows"))]
+	#[test]
+	fn test_readonly_dir() {
+		let persister = FilesystemPersister::new("test_readonly_dir_persister".to_string());
+		let test_writeable = TestWriteable{};
+		let test_txo = OutPoint {
+			txid: Txid::from_hex("8984484a580b825b9972d7adb15050b3ab624ccd731946b3eeddb92f4e7ef6be").unwrap(),
+			index: 0
+		};
+		// Create the persister's directory and set it to read-only.
+		let path = &persister.path_to_channel_data;
+		fs::create_dir_all(path).unwrap();
+		let mut perms = fs::metadata(path).unwrap().permissions();
+		perms.set_readonly(true);
+		fs::set_permissions(path, perms).unwrap();
+		match persister.write_channel_data(test_txo, &test_writeable) {
+			Err(e) => {
+				assert_eq!(e.kind(), io::ErrorKind::PermissionDenied);
+			}
+			_ => panic!("Unexpected error message")
+		}
+	}
+
+	// Test failure to rename in the process of atomically creating a channel
+	// monitor's file. We induce this failure by making the `tmp` file a
+	// directory.
+	// Explanation: given "from" = the file being renamed, "to" = the
+	// renamee that already exists: Windows should fail because it'll fail
+	// whenever "to" is a directory, and Unix should fail because if "from" is a
+	// file, then "to" is also required to be a file.
```

Comment:
Doesn't this actually test temp file creation (line 81) rather than the rename (line 85)? It's a good test to have regardless, though I think in both cases creating a read-only version of the temp and final files, respectively, may be sufficient to handle any platform... shooting from the hip on the last part.

Reply:
> I think in both cases creating a read-only version of the temp and final files, respectively, may be sufficient to handle any platform 

I wasn't sure what this part meant. 

(Btw I remembered last night you're on vacation right now, I'm just responding to your responses for posterity). But, good catch and I kept this test and added an actual rename faliure unit test.

Reply:
Was just thinking for either test, the presence of a read-only file ("to" or "from" depending on the test) might be sufficient to trigger the case without using a directory. Using a directory is an extra detail for the reader (different type vs invalid permission), so if possible it may be better to avoid.

Reply:
That doesn't seem work, at least on unix. This is what I tried changing the test to: https://i.imgur.com/yOHqQB5.png (sorry not the most readabl). Lmk if i'm missing something